### DISCUSSION TITLE
request: Rename the float types to mark the backward compatibility to C like "f32c" and "f64c", and add some new ones with the old names and correct math

### DISCUSSION THREAD

**Discussion by debilosaurus** [vlang/v#18613] [Ideas]
playground

```
println("f32: 0.1\*0.5 = "+(f32(0.1)*f32(0.5)).str()+" -> okay")
println("f32: 0.1\*0.1 = "+(f32(0.1)*f32(0.1)).str()+" -> why so much zeroes?")
println("f64: 0.1\*0.1 = "+(f64(0.1)*f64(0.1)).str()+" -> i dont trust you ...")
```

**What did you expect to see?**
I am no computer scientist, so maybe this values make sense in some way,
but I expected somehow mathematical correct values.
 
**What did you see instead?**
```
f32: 0.1*0.5 = 0.05 -> okay
f32: 0.1*0.1 = 0.01000000 -> why so much zeroes?
f64: 0.1*0.1 = 0.010000000000000002 -> i dont trust you ...
```

**Comment by JalonSolov** [comment]
The numbers can be explained fairly easily by knowing that floating point is implemented to the IEEE-754 standard... which can show round-off errors, etc., such as shown in your example.

Your example can also be simplified a bit...
```v
println('f32: 0.1*0.5 = ${f32(0.1)*f32(0.5)} -> okay')
println('f32: 0.1*0.1 = ${f32(0.1)*f32(0.1)} -> why so much zeroes?')
println('f64: 0.1*0.1 = ${f64(0.1)*f64(0.1)} -> i dont trust you ...')
```
IEEE-754 standard is the most common floating point implementation, used by V, C, and most other programming languages, as well as most floating point hardware.

**Comment by medvednikov** [comment]
In Go it's handled nicely:

https://go.dev/play/p/2pdcG5LbfiB

```
f32: 0.1*0.5 = 0.050000 
f32: 0.1*0.1 = 0.010000  
f64: 0.1*0.1 = 0.010000  
```

**Comment by JalonSolov** [comment]
Yep... because they force rounding, etc.  V doesn't.  There are lib routines to do it if you want.

**Comment by debilosaurus** [comment]
Do I get you right, and I hope you dont mind if this sounds a little bit disappointed
1. the purity of some IEEE standard is more important than the correctness of basic arithmetics, and
2. the suggested workaround is that every programmer should consider possible wrong results and fix them.

Wouldn't it be simpler and much more transparent the other way round:
- if the arithmetic operators would work like they should, and
- there are some low-level functions defined in a separate modul like "math.ieee754.add()" for the other cases
(eg. compatibility with old software, which expects and fixes this behaviour),

**Comment by JalonSolov** [comment]
It would actually be more preferable if correctness were the main goal, except that it would be completely against the way almost everything else works.

For example, if you translate C code to V code, you could wind up with different results between the 2 versions, even though the code "looks" correct, because V would doing the math more correct than the C code, without changing anything.

Otherwise, during your conversion, you couldn't just keep `0.1 * 0.1`, you would have to translate that to `math.ieee754.mult(0.1, 0.1)` or similar, making translating code much harder.

Perhaps when there is as much V code as there is C code, this would be acceptable.  Right now, it actually makes sense to support ease of translation, while allowing users to have the more correct values _if_ they decide they want them.

**Comment by debilosaurus** [comment]
I had time to think about it. So pls can you convert this issue to an feature request:

1. Request: Rename the float types to mark the backward compatibility to C like "f32c" and "f64c", and add some new ones with the old names and correct math.

Because I really like the new approaches of this language, here - as my conclusion - some other suggestions, which may result in better first user experience. Mine was every "mixed". I stumbled from one compiler errror into the next, doing strange fixes of things the language should take care of:

2. Pls fix the issues around the arrays with constant length and typing (nesting, initialisation). It is nearly impossible to generate readable code without converting this into structs and generate own "new" functions for return values. If this structs are needed the language should generate it, and if there must be some kind of a reference to get them back as a return value, there should be chance to do it without to much efford. Simple code never worked: "type Dot = [3]f64 ; type Box = [2]Dot ; a:= <init?>, println(a)"
3. After a while I was totaly confused about the brackets type, and I tried every type to get rid of the type and initialization errors. I am not shure about this point, but I never had any problems like this in C/Python/Go.
4. Very annoying is the point, that it seems to be impossible to define a struct in one line and to set default values for arguments.
5. The keyword "mut" isnt consistent too - sometimes as "mut:", most of the time only "mut". Together with (4.) I would like to see that there is no "mut:" for more than one line. Every mutable should have to be prefixed.
6. I dont like the idea of importing modules without a name space. This should be the exception (enforced by some keyword) and not the default behaviour of eg. selective imports.
7. Most important: Pls rethink the decision to avoid Nim's killer feature - always using the first argument for "dot-piping" objects. I understand the point of view with "only one solution for each problem", but this is such a migthy simplification: no need for the bracket in front of the function name. I am shure this would be worth it - in some cases less restriction (or more options) will give the room, that something like a coding style can evolve.

**Comment by spytheman** [comment]
> Yep... because they force rounding, etc. V doesn't. There are lib routines to do it if you want.

There is also the builtin string interpolation formatting:
```v
println('f32: 0.1*0.5 = ${f32(0.1)*f32(0.5):0.6f}')
println('f32: 0.1*0.1 = ${f32(0.1)*f32(0.1):0.6f}')
println('f64: 0.1*0.1 = ${f64(0.1)*f64(0.1):0.6f}')
```
which produces:
```
f32: 0.1*0.5 = 0.050000
f32: 0.1*0.1 = 0.010000
f64: 0.1*0.1 = 0.010000
```
which is the same output as the Go one.

**Comment by debilosaurus** [comment]
Thank you very much for your replies, and I am sorry that I couldnt answer earlier.
Maybe a precision module (v has its advantages) would be another option to think about:

```
import math

struct F64p {
    f f64
    p f32
}

fn new_f64p(r f64) F64p {
    return F64p{f:r, p:f32(r*math.pow(2,-52))}
}
fn (l F64p) f64() f64 {
    d := 0-int(math.ceil(math.log10(l.p)))
    return math.round_sig(l.f, d)
}
fn (l F64p) str() string {
    return l.f64().str()
}

fn (l F64p) +(r F64p) F64p {
    return F64p{f:l.f+r.f, p:l.p+r.p}
}
fn (l F64p) -(r F64p) F64p {
    return F64p{f:l.f-r.f, p:l.p+r.p}
}
fn (l F64p) *(r F64p) F64p {
    return F64p{f:l.f*r.f, p:f32(l.f*r.p+r.f*l.p)}
}
fn (l F64p) /(r F64p) F64p {
    return F64p{f:l.f/r.f, p:f32((l.f*r.p+r.f*l.p)/r.f/r.f)}
}

fn main() {
    f1, f2 := 0.1, 0.35
    p1, p2 := new_f64p(f1), new_f64p(f2)
    println("${f1+f2} vs. ${p1+p2}")
    println("${f1-f2} vs. ${p1-p2}")
    println("${f1*f2} vs. ${p1*p2}")
    println("${f1/f2} vs. ${p1/p2}")
}
```
